### PR TITLE
impl(GCS+gRPC): add `*Connection` class for async client

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #568

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9120)

### CODE REVIEW

File: google/cloud/storage/internal/async_connection_impl.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+class AsyncConnectionImpl : public AsyncConnection {
+ public:
+  explicit AsyncConnectionImpl(CompletionQueue cq,
+                               std::shared_ptr<StorageStub> stub,
+                               Options options);
+  ~AsyncConnectionImpl() override = default;
+
+  Options options() const override { return options_; }
```
Comment (dbolduc) [inline google/cloud/storage/internal/async_connection_impl.h:40]: do you want a test for this? maybe that one option from `DefaultOptionGrpc()` is set?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/async_connection_impl.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  future<Status> AsyncDeleteObject(
+      storage::internal::DeleteObjectRequest const& request) override;
+
+ private:
+  std::unique_ptr<storage::RetryPolicy> retry_policy() {
+    auto const& options = internal::CurrentOptions();
+    if (options.has<storage::RetryPolicyOption>()) {
```
Comment (dbolduc) [inline google/cloud/storage/internal/async_connection_impl.h:48]: ugh, forgot to add this comment to my review.

The current options vs connection options behavior isn't tested. I decided to test it in bigtable, but [it was annoying](https://github.com/googleapis/google-cloud-cpp/blob/9c6256e77b03e83cee73faea2cf4880f807612df/google/cloud/bigtable/internal/data_connection_impl_test.cc#L113-L182). Your call on whether or not to add a test.

File: google/cloud/storage/internal/async_connection_impl.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/async_connection_impl.h:40]: We check that `AuthorityOption` is set to the right value.  Is that what you meant?  I included a comment to explain that too.

File: google/cloud/storage/internal/async_connection_impl_test.cc [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                        UnorderedElementsAre(
+                            Pair("x-goog-quota-user", "test-quota-user"),
+                            Pair("x-goog-fieldmask", "field1,field2")));
+            EXPECT_THAT(request.bucket(), "projects/_/buckets/test-bucket");
+            EXPECT_THAT(request.object(), "test-object");
+            return make_ready_future(PermanentError());
+          });
+  CompletionQueue cq;
+  auto connection = MakeTestConnection(cq, mock);
+  OptionsSpan span(connection->options());
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/storage/internal/async_connection_impl_test.cc:83]: Ah. I think this is kind of a problem we will have to deal with eventually. I added a discussion item for the next meeting.

It's probably cleanest to just:
```c++
ConnectionImpl::Foo(FooRequest const& r) {
  auto options = internal::MergeOptions(internal::CurrentOptions(), options_);
  OptionsSpan span(options);
  
  return internal::RetryLoop(
    options.get<RetryPolicyOption>()->clone(),
    options.get<BackoffPolicyOption>()->clone(),
    options.get<IdempotencyPolicyOption>()->clone()->Foo(r),
    /* etc */ );
}
```
and drop the [policy getters](https://github.com/googleapis/google-cloud-cpp/blob/5f84aede7a748d723cfbabe979f376925b37127c/google/cloud/accessapproval/internal/access_approval_connection_impl.h#L87-L120) altogether.

**Pull request merged**